### PR TITLE
Switch from gnuindent to clang-format

### PR DESCRIPTION
**Pull request by sandro-elsweijer**
**_Describe your changes here:_**
Switch from `gnuindent` to `clang-format`, as described in #12.
The indentation from `gnuindent` is not perfectly replicable with `clang-format`, but I think I am as close as I can get.

There are still some options to experiment with (especially the C++ related options), so that our code looks nice. I did not indent the source code yet, because we are waiting for the updates for version 1.0 to be merged, so we can avoid as many merge conflicts as possible.
Feel free to look if the formatting looks good for you, I am happy over each suggestion.

The biggest changes are:
1. `gnuindent` aligns consecutive declarations like this:  
```c
int         var1;
char        var2;
longlong    var3;
```  
`clang-format` has the option to do this as well, but then it also aligns the declarations inside function definitions:  
```c
void myfunc (int         var1,
             char        var2,
             longlong    var3);
```
Therefore, the alignment of consecutive declarations is disabled.

2. The column limit is now 120 instead of 80
3. `gnuindent` sometimes breaks before `&&`, `||` etc. and sometimes after them. `clang-format` now always breaks before `&&`, `||` etc.
4. `/* *INDENT-OFF* */` has to be switched to `/* clang-format off */`. I have found no way to edit the clang macro.
5. C++ gets indented differently by design. This is the reason why we want to move away from `gnuindent`. 
6. `gnuindent` ignored the length of comments, `clang-format` does not. Therefore, long, but still nice looking lines from `gnuindent`  are now broken, because we often use trailing comments instead of leading comments.

**_All these boxes must be checked by the reviewers before merging the pull request:_**

- [ ] The author added a BSD statement to `doc/` (or already has one)
- [ ] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)

  If the Pull request indroduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

- [ ] All tests pass (in various configurations, this should be executed automatically in a github action)
- [ ] New source/header files are properly added to the Makefiles
- [ ] The reviewer executed the new code features at least once and checked the results manually
- [ ] The code is covered in an existing or new test case
- [ ] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [ ] The code is well documented
- [ ] All function declarations, structs/classes and their members have a proper doxygen documentation
- [ ] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)
- [ ] Testing of this template: If you feel something is missing from this list, contact the developers

### CODE REVIEW

**Comment by holke** [comment]
When executing the `check_if_file_indendet` script my clang-format complains:
```
clang-format: Unknown command line argument '--dry-run'.  Try: '/usr/bin/clang-format -help'
clang-format: Did you mean '-icp-lto'?
clang-format: Unknown command line argument '--Werror'.  Try: '/usr/bin/clang-format -help'
clang-format: Did you mean '-cursor'?
```

i am running clang-format version 6.0.0:
```
$ clang-format --version
clang-format version 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
```

If these arguments are needed we need to specify which clang-format version is needed and check that we fulfill before trying to indent the files.

**Comment by holke** [comment]
clang-format-10 seems to be the first fully supported version.
It is available on my ubuntu, but i have to explicitly specify clang-format-10 in the indent scripts.

**Comment by holke** [comment]
Do you know what causes this behaviour:
```
-  else {                        /* set_from != NULL */
-    t8_forest_t         forest_from = forest->set_from; /* temporarily store set_from, since we may overwrite it */
+  else { /* set_from != NULL */
+    t8_forest_t forest_from
+      = forest
+          ->set_from; /* temporarily store set_from, since we may overwrite it */
```

The new variant looks rather ugly tbh.

File: .clang-format [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  BeforeElse:      true
+  IndentBraces:    false
+  SplitEmptyFunction: true
+  SplitEmptyRecord: true
+  SplitEmptyNamespace: true
+BreakBeforeBinaryOperators: All
+BreakBeforeBraces: Custom
+BreakBeforeTernaryOperators: 'true'
+BreakInheritanceList: AfterColon
+ColumnLimit: '80'
```
Comment (holke) [inline .clang-format:41]: Shouldn't this by 120?

**Comment by holke** [comment]
Thanks for adding this and putting the work into it.
We definitely need this change.

I am playing around with it a bit and add a few comments regarding fine tuning.

File: .clang-format (hunk 1 above)
Comment (Davknapp) [reply .clang-format]: Changed

**Comment by Davknapp** [comment]
> Do you know what causes this behaviour:
> 
> ```
> -  else {                        /* set_from != NULL */
> -    t8_forest_t         forest_from = forest->set_from; /* temporarily store set_from, since we may overwrite it */
> +  else { /* set_from != NULL */
> +    t8_forest_t forest_from
> +      = forest
> +          ->set_from; /* temporarily store set_from, since we may overwrite it */
> ```
> 
> The new variant looks rather ugly tbh.

I am unable to reproduce this behaviour using clang-format-16

**Comment by Davknapp** [comment]
> > Do you know what causes this behaviour:
> > ```
> > -  else {                        /* set_from != NULL */
> > -    t8_forest_t         forest_from = forest->set_from; /* temporarily store set_from, since we may overwrite it */
> > +  else { /* set_from != NULL */
> > +    t8_forest_t forest_from
> > +      = forest
> > +          ->set_from; /* temporarily store set_from, since we may overwrite it */
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The new variant looks rather ugly tbh.
> 
> I am unable to reproduce this behaviour using clang-format-16

Retried with version 10.0.1 and can't reproduce this error either.

**Review by Davknapp — APPROVED** [review]

**Pull request merged**